### PR TITLE
Move WatchCache to pkg/storage and make it private

### PR DESCRIPTION
**Pull request by wojtek-t**
This is a follow-up PR from discussion in #12721

The purpose of this PR is:
(1) change the reflector.go so that it doesn't have any reference to WatchCache
(2) make WatchCache to be a private class in pkg/storage (in fact it's just implementation detail)

However to do (1) I need to change the signature of Store::Replace() function to take resourceVersion as an additional argument (we need it for correctness in WatchCache).
I think this change is fine, because we use the store implementations together with Reflector (which already has a notion of resourceVersion) - the only places where we need to explicitly pass the some resource version are tests - I didn't have to modify any "production" code.

@bprashanth @timothysc

### CODE REVIEW

**Comment by bprashanth** [comment]
Can you justify this in the pr description? I guess I'm missing some context. I prefer a local hack to a global change at the interface level. Something like thread safe store should really not have to care about resource version, it should just be a locked hash map. Everyone who wants to use a locked cached would have to deal with this etcd style vector clock notation.

**Comment by wojtek-t** [comment]
The purpose of this PR is:
(1) change the reflector.go so that it doesn't have any reference to WatchCache
(2) make WatchCache to be a private class in pkg/storage (in fact it's just implementation detail)

However to do (1) I need to change the signature of Store::Replace() function to take resourceVersion as an additional argument (we need it for correctness in WatchCache).
I think this change is fine, because we use the store implementations together with Reflector (which already has a notion of resourceVersion) - the only places where we need to explicitly pass the some resource version are tests - I didn't have to modify any "production" code.

**Comment by wojtek-t** [comment]
@bprashanth - I extended the PR description with explanation of what I'm doing. 

PTAL

**Comment by wojtek-t** [comment]
@bprashanth - friendly ping - PTAL

cc @lavalamp [for opinion about it]

**Comment by bprashanth** [comment]
> (1) change the reflector.go so that it doesn't have any reference to WatchCache

Why is this important? To do it, your're changing something that has nothing to do with the reflector or watch cache, to understand a watch concept. I want to be able to use something like thread-safe-store from the kubelet or networking package as a generic map with locks. Piping through a resource version for replace is not horrible, but doesn't make sense in that context.  I'd rather avoid it, but I'm not hard against it.

**Comment by wojtek-t** [comment]
Thanks @bprashanth - I get your point (although I'm not fully convinced).

@lavalamp - can you please make the final call on that?

File: pkg/client/unversioned/cache/fifo_test.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func TestFIFO_addUpdate(t *testing.T) {
 func TestFIFO_addReplace(t *testing.T) {
 	f := NewFIFO(testFifoObjectKeyFunc)
 	f.Add(mkFifoObj("foo", 10))
-	f.Replace([]interface{}{mkFifoObj("foo", 15)})
+	f.Replace([]interface{}{mkFifoObj("foo", 15)}, "0")
```
Comment (lavalamp) [inline pkg/client/unversioned/cache/fifo_test.go]: Nit: Real data probably would have rv being >= the highest RV in the list. So if that 15 is supposed to be a resource version, then the "0" should probably be at least "15"...

**Comment by lavalamp** [comment]
@bprashanth @wojtek-t I think we want to evolve these caches into things that understand our objects, eventually into write-through caches etc. So I think being more explicit about resource versions in these caches seems like a good change to me.

**Comment by wojtek-t** [comment]
I think I agree with @lavalamp - my thinking is that in the fullness of time all those caches will be caches for "runtime.Object", which is already some dependence on resource versions. So I think it makes sense to proceed with that PR.
I will rebase it today.

**Comment by wojtek-t** [comment]
@bprashanth @lavalamp - PTAL

**Comment by wojtek-t** [comment]
@lavalamp / @bprashanth - friendly ping

File: pkg/client/unversioned/cache/poller.go [hunk 2]
```diff
@@ -1,86 +0,0 @@
-/*
```
Comment (lavalamp) [inline pkg/client/unversioned/cache/poller.go]: Why is this removal part of this change?

File: pkg/client/unversioned/cache/reflector_test.go [hunk 3]
```diff
@@ -359,34 +359,3 @@ func TestReflector_ListAndWatchWithErrors(t *testing.T) {
 		r.ListAndWatch(util.NeverStop)
 	}
 }
-
-func TestReflectorForWatchCache(t *testing.T) {
```
Comment (lavalamp) [inline pkg/client/unversioned/cache/reflector_test.go]: Why is this test removed?

Comment (lavalamp) [inline pkg/client/unversioned/cache/reflector_test.go]: Oh, I see, it was moved. Nvm

**Comment by lavalamp** [comment]
I am fine with this, just a question + nit

File: pkg/client/unversioned/cache/poller.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/client/unversioned/cache/poller.go]: The poller is not used anywhere in the code.
However, it uses Replace method, which I would need to update, so I decided to just remove it.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: pkg/client/unversioned/cache/fifo_test.go (hunk 1 above)
Comment (lavalamp) [inline pkg/client/unversioned/cache/fifo_test.go]: Missing push?

**Comment by wojtek-t** [comment]
@lavalamp - friendly ping

File: pkg/client/unversioned/cache/fifo_test.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/client/unversioned/cache/fifo_test.go]: yup - sorry. Should be fixed now.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

**Comment by wojtek-t** [comment]
@lavalamp - friendly ping :)

**Comment by wojtek-t** [comment]
@lavalamp - friendly ping :)

**Comment by lavalamp** [comment]
LGTM

**Comment by lavalamp** [comment]
Sorry for delay!

**Comment by wojtek-t** [comment]
Test failure is a flake:
"Services should serve a basic endpoint from pods"

@k8s-bot test this please

**Comment by wojtek-t** [comment]
Next flake:
Kubernetes e2e suite.Services should serve multiport endpoints from pods
Kubernetes e2e suite.Services should be able to up and down services

@k8s-bot test this please

**Pull request merged**